crates/store/re_chunk/tests/memory_test.rs

```diff
@@ -0,0 +1,193 @@
+                    ArrowListArray::<i32>::default_datatype(scalars.data_type().clone()),
+                    ArrowOffsets::try_from_lengths(
+                        std::iter::repeat(NUM_SCALARS as usize / 10).take(10),
+                    )
+                    .unwrap()
+                    .into(),
+                    scalars.to_boxed(),
+                    None,
+                )
+            });
+
+            let filter = ArrowBooleanArray::from_slice(
+                (0..unfiltered.0.len()).map(|i| i % 2 == 0).collect_vec(),
+            );
+            let filtered = memory_use(|| re_chunk::util::filter_array(&unfiltered.0, &filter));
+
+            (unfiltered, filtered)
+        });
+
+    eprintln!(
+        "unfiltered={} filtered={} total={}",
+        re_format::format_bytes(unfiltered_size_bytes as _),
+        re_format::format_bytes(filtered_size_bytes as _),
+        re_format::format_bytes(total_size_bytes as _),
+    );
+
+    assert!(unfiltered_size_bytes + filtered_size_bytes <= total_size_bytes);
+    assert!(unfiltered_size_bytes <= filtered_size_bytes * 2);
+
+    {
+        let unfiltered = unfiltered
+            .values()
+            .as_any()
+            .downcast_ref::<ArrowPrimitiveArray<i64>>()
+            .unwrap();
+        let filtered = filtered
+            .values()
+            .as_any()
+            .downcast_ref::<ArrowPrimitiveArray<i64>>()
+            .unwrap();
+
+        assert!(
+            !std::ptr::eq(
+                unfiltered.values().as_ptr_range().start,
+                filtered.values().as_ptr_range().start
+            ),
+            "data should be copied -- pointers shouldn't match"
+        );
+    }
+}
+
+#[test]
+// TODO(cmc): That's the end goal, but it is simply impossible with `ListArray`'s encoding.
+//            See `Chunk::take_array`'s doc-comment for more information.
+#[should_panic = "assertion failed: untaken_size_bytes > taken_size_bytes * 10"]
+fn take_does_not_allocate() {
+    re_log::setup_logging();
+
+    const NUM_SCALARS: i64 = 10_000_000;
+
+    let (((untaken, untaken_size_bytes), (taken, taken_size_bytes)), total_size_bytes) =
+        memory_use(|| {
+            let untaken = memory_use(|| {
+                let scalars = ArrowPrimitiveArray::from_vec((0..NUM_SCALARS).collect_vec());
+                ArrowListArray::<i32>::new(
+                    ArrowListArray::<i32>::default_datatype(scalars.data_type().clone()),
+                    ArrowOffsets::try_from_lengths(
+                        std::iter::repeat(NUM_SCALARS as usize / 10).take(10),
+                    )
+                    .unwrap()
+                    .into(),
+                    scalars.to_boxed(),
+                    None,
+                )
+            });
+
+            let indices = ArrowPrimitiveArray::from_vec(
+                (0..untaken.0.len() as i32)
+                    .filter(|i| i % 2 == 0)
+                    .collect_vec(),
+            );
```

Comment:
What happens if we take a contiguous slice? Is there internal optimization for slicing when possible, or does it always copy blindly?

Reply:
Always copies.